metaflow/plugins/kfp/kfp.py

```diff
@@ -1,112 +1,334 @@
+from .constants import S3_AWS_ARN as S3_AWS_ARN_VALUE
+from .constants import S3_BUCKET as S3_BUCKET_VALUE
+from typing import NamedTuple
+from collections import deque
 
-def get_ordered_steps(graph):
+StepOutput = NamedTuple('StepOutput', [('ds_root', str), ('run_id', str)])
+
+def step_op_func(python_cmd_template, step_name: str,
+                 code_url: str,
+                 ds_root: str,
+                 run_id: str,
+               ):
     """
-    Returns the ordered step names in the graph (FlowGraph) from start step to end step as a list of strings containing the
-    step names.
+    Function used to create a KFP container op (see: `step_container_op`) that corresponds to a single step in the flow.
 
-    # TODO: Support other Metaflow graphs, as branching and joins are not currently supported
-    Note: All MF graphs start at the "start" step and end with the "end" step (both of which are mandatory).
     """
+    import subprocess
+    import os
+
+    MODIFIED_METAFLOW_URL = 'git+https://github.com/zillow/metaflow.git@branch-and-join'
+    DEFAULT_DOWNLOADED_FLOW_FILENAME = 'downloaded_flow.py'
+
+    print("\n----------RUNNING: CODE DOWNLOAD from URL---------")
+    subprocess.call(
+        ["curl -o {downloaded_file_name} {code_url}".format(downloaded_file_name=DEFAULT_DOWNLOADED_FLOW_FILENAME,
+                                                            code_url=code_url)], shell=True)
+
+    print("\n----------RUNNING: KFP Installation---------------")
+    subprocess.call(["pip3 install kfp"], shell=True)  # TODO: Remove this once KFP is added to dependencies
+
+    print("\n----------RUNNING: METAFLOW INSTALLATION----------")
+    subprocess.call(["pip3 install --user --upgrade {modified_metaflow_git_url}".format(
+        modified_metaflow_git_url=MODIFIED_METAFLOW_URL)],
+                    shell=True)
+
+    print("\n----------RUNNING: MAIN STEP COMMAND--------------")
+    S3_BUCKET = os.getenv("S3_BUCKET")
+    S3_AWS_ARN = os.getenv("S3_AWS_ARN")
+
+    define_s3_env_vars = 'export METAFLOW_DATASTORE_SYSROOT_S3="{}" && export METAFLOW_AWS_ARN="{}" '.format(S3_BUCKET,
+                                                                                                             S3_AWS_ARN)
+    define_username = 'export USERNAME="kfp-user"'
+    python_cmd = python_cmd_template.format(ds_root=ds_root, run_id=run_id)
+
+    final_run_cmd = f'{define_username} && {define_s3_env_vars} && {python_cmd}'
+
+    print("RUNNING COMMAND: ", final_run_cmd)
+    proc = subprocess.run(final_run_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
+    proc_output = proc.stdout
+    proc_error = proc.stderr
+
+    # END is the final step and no outputs need to be returned
+    if step_name.lower() == 'end':
+        print("_______________ FLOW RUN COMPLETE ________________")
+
+    # TODO: Check why MF echo statements are getting redirected to stderr
```

Comment:
Not sure if this is related, so take this with a grain of salt, but potentially it's the result of python logging to `stderr` by default, see [here](https://stackoverflow.com/a/58971396/724251). Essentially all messages outputted via the `logging` module get outputted to `stderr` by default as they are considered "internal" messages.

It has caught me out plenty of times in my career thus far as my brain computes that `logging.info` (and above) should go to `stdout` but that's not the case!

Reply:
Thank you for sharing @alexlatchford !

Reply:
Thank you. I didn't know this! 

I just looked at why this is happening and in this case it turns out it's not to do with the logging but to do with the fact that Metaflow code which uses click's `secho` is defaulting to using `err=True` inside the [`echo always`](https://github.com/zillow/metaflow/blob/master/metaflow/cli.py#L54) function that gets called. And, that seems to be [the reason](https://click.palletsprojects.com/en/7.x/api/#click.echo) why it is getting directed to stderr instead of stdout.

Reply:
Now fixed it to use stdout instead where required